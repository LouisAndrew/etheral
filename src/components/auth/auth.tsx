import React, { useState, useEffect } from 'react';
import { useNavigate } from '@reach/router';
import { useDispatch } from 'react-redux';

import { Flex, Box, Button } from 'rebass';

import Login from './login';
import SignUp from './signup';
import useAllProducts from 'helper/use-all-products';
import { FirebaseUserData, InCart } from 'helper/schema/firebase-user';
import { setCart } from 'state/actions/cart';
import extractCartFirestore from 'helper/extract-cart-firestore';

type Props = {
    auth: firebase.auth.Auth;
    db: firebase.firestore.Firestore;
    googleProvider: firebase.auth.GoogleAuthProvider;
    authPersistence: string;
};

export type LoginProps = {
    email: string;
    password: string;
};

export type SignUpProps = LoginProps & {
    name: string;
    signupNewsletter: boolean;
};

// enum to show where the error occurs.
enum ErrorTarget {
    NONE,
    LOGIN,
    SIGN_UP,
}

const Auth: React.FC<Props> = ({
    auth,
    db,
    googleProvider,
    authPersistence,
}) => {
    const [uid, setUid] = useState('');
    const [isNewUser, setIsNewUser] = useState(false);

    // state to store errror msg.
    const [error, setError] = useState<firebase.auth.Error | undefined>(
        undefined
    );
    const [errTarget, setErrTarget] = useState<ErrorTarget>(ErrorTarget.NONE);

    const navigate = useNavigate();
    const allProducts = useAllProducts();
    const dispatch = useDispatch();

    // set auth persistence.
    useEffect(() => {
        auth.setPersistence(authPersistence);
    }, []);

    useEffect(() => {
        // call this function just if user is logged in!
        const currentUser = auth.currentUser;
        if (currentUser !== null) {
            (async () => {
                if (!uid) {
                    await setUid(currentUser.uid);
                }

                if (!isNewUser) {
                    await fetchCartItems(currentUser);
                } else {
                    await createNewUser(currentUser);
                }

                await navigate('/');
            })();
        }
    }, [uid]);

    const createNewUser = async (currentUser: firebase.User) => {
        if (currentUser !== null) {
            try {
                const userData: FirebaseUserData = {
                    name: currentUser.displayName
                        ? currentUser.displayName
                        : '',
                    email: currentUser.email ? currentUser.email : '',
                    inCart: [],
                    wishlist: [],
                    orders: [],
                };

                await db
                    .collection('user')
                    .doc(currentUser.uid)
                    .set(userData);
            } catch (e) {
                console.error(e);
            }
        }
    };

    const fetchCartItems = async (currentUser: firebase.User) => {
        // fetch items on cart and wishlist of the user => on login.
        if (currentUser) {
            try {
                const docRef = await db
                    .collection('user')
                    .doc(currentUser.uid)
                    .get();

                const userData = await docRef.data();

                // fetch and checks for all undefined object.
                if ((await docRef.exists) && userData) {
                    const inCart = (await (userData.inCart as any)) as InCart;
                    const onWishlist = (await (userData.wishlist as any)) as InCart;

                    const filteredInCartData = await extractCartFirestore({
                        firestoreCartData: inCart,
                        allProducts,
                    });
                    const filteredWishlistData = await extractCartFirestore({
                        firestoreCartData: onWishlist,
                        allProducts,
                    });

                    await dispatch(
                        setCart({
                            cart: filteredInCartData,
                            wishlist: filteredWishlistData,
                        })
                    );
                }
            } catch (e) {
                console.error(e);
            }
        }
    };

    // user login
    const loginWithEmail = async ({ password, email }: LoginProps) => {
        try {
            const loginReq = await auth.signInWithEmailAndPassword(
                email,
                password
            );
            const { user } = await loginReq;

            if (user !== null) {
                setUid(user.uid);
            }
        } catch (e) {
            setErrTarget(ErrorTarget.LOGIN);
            setError(e);
        }
    };

    // create new user
    const signupWithEmail = async ({
        name,
        password,
        email,
        signupNewsletter,
    }: SignUpProps) => {
        try {
            const signupReq = await auth.createUserWithEmailAndPassword(
                email,
                password
            );

            const { user } = await signupReq;

            if (user !== null) {
                await user.updateProfile({
                    displayName: name,
                });
                await setIsNewUser(true);
                await setUid(user.uid);
            }

            if (await signupNewsletter) {
                // call cloud function to sign this acc to newsletter up
                console.log('Do signup newsletter here');
            }
        } catch (e) {
            setErrTarget(ErrorTarget.SIGN_UP);
            setError(e);
        }
    };

    // sign in/up with google
    const withGoogle = async () => {
        try {
            const signinReq = await auth.signInWithPopup(googleProvider);

            const { user } = await signinReq;

            if (user !== null) {
                if (signinReq.additionalUserInfo) {
                    await setIsNewUser(signinReq.additionalUserInfo.isNewUser);
                }
                await setUid(user.uid);
            }
        } catch (e) {
            console.error(e);
        }
    };

    return (
        <Flex
            width={['100%', '100%', '100%', '70%', '60%']}
            justifyContent="space-between"
            flexWrap="wrap"
            px={[4, 4, 6]}
            m="0 auto"
        >
            <Box width={['100%', '100%', '48%', '45%', '44%']} my={[5]}>
                <SignUp
                    signup={signupWithEmail}
                    firebaseError={
                        errTarget === ErrorTarget.SIGN_UP ? error : undefined
                    }
                />
            </Box>
            <Box
                width={['100%', '100%', '48%', '45%', '44%']}
                my={[5]}
                sx={{ textAlign: 'center' }}
            >
                <Login
                    login={loginWithEmail}
                    withGoogle={withGoogle}
                    firebaseError={
                        errTarget === ErrorTarget.LOGIN ? error : undefined
                    }
                />
            </Box>
        </Flex>
    );
};

export { Auth };
