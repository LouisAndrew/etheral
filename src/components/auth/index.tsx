import React from 'react';
import firebase from 'gatsby-plugin-firebase';

import { Auth as AuthEl } from './auth';

const Auth = <AuthEl auth={firebase.auth()} db={firebase.firestore()} />;

export default Auth;
