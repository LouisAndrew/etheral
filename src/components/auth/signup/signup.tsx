import React from 'react';
import { useForm, FormProvider } from 'react-hook-form';
import { debounce } from 'lodash';

import { Box, Heading, Text, Button } from 'rebass';
import { Input, Label } from '@rebass/forms';

import { SignUpProps } from '../auth';
import Checkbox from 'components/checkbox';

type Props = {
    firebaseError?: firebase.auth.Error | undefined;
    signup: (args: SignUpProps) => void;
};

type Inputs = {
    signupFirstName: string;
    signupLastName: string;
    signupEmail: string;
    signupPassword: string;
    signupConfirmPassword: string;
    signupNewsletter: boolean;
};

const SignUp: React.FC<Props> = ({ firebaseError, signup }) => {
    const methods = useForm<Inputs>();
    const { register, handleSubmit, getValues, errors } = methods;

    const debouncedSignup = debounce(signup, 500);

    const submit = ({
        signupEmail,
        signupFirstName,
        signupLastName,
        signupPassword,
        signupNewsletter,
    }: Inputs) => {
        const name = `${signupFirstName} ${signupLastName}`;

        debouncedSignup({
            name,
            password: signupPassword,
            email: signupEmail,
            signupNewsletter,
        });
    };

    const textAlignAttr: 'left' | 'center' = 'left';

    const errorStyling = {
        variant: 'formError',
        textAlign: textAlignAttr,
        mt: [3],
        role: 'alert',
    };

    return (
        <FormProvider {...methods}>
            <Box
                as="form"
                onSubmit={handleSubmit(submit)}
                sx={{ textAlign: 'center' }}
            >
                <Heading as="h3" variant="formHeading" mb={[4]}>
                    NEW ACCOUNT
                </Heading>

                <Box>
                    <Label variant="text.formLabel" htmlFor="signup-first-name">
                        FIRST NAME
                    </Label>
                    <Input
                        id="signup-first-name"
                        name="signupFirstName"
                        type="text"
                        variant="variants.authInput"
                        ref={register({
                            required: true,
                        })}
                    />
                    {errors.signupFirstName && (
                        <Text {...errorStyling}>Please enter your name</Text>
                    )}

                    <Label variant="text.formLabel" htmlFor="signup-last-name">
                        LAST NAME
                    </Label>
                    <Input
                        id="signup-last-name"
                        name="signupLastName"
                        type="text"
                        variant="variants.authInput"
                        ref={register}
                    />

                    <Label variant="text.formLabel" htmlFor="signup-email">
                        EMAIL
                    </Label>
                    <Input
                        name="signupEmail"
                        id="signup-email"
                        type="email"
                        variant="variants.authInput"
                        ref={register({
                            required: true,
                        })}
                    />
                    {errors.signupEmail && (
                        <Text {...errorStyling}>Please enter your email</Text>
                    )}
                    {firebaseError &&
                    firebaseError.code === 'auth/email-already-in-use' ? (
                        <Text {...errorStyling}>{firebaseError.message}</Text>
                    ) : null}

                    <Label variant="text.formLabel" htmlFor="signup-password">
                        PASSWORD
                    </Label>
                    <Input
                        name="signupPassword"
                        id="signup-password"
                        type="password"
                        variant="variants.authInput"
                        ref={register({
                            required: true,
                        })}
                    />
                    {errors.signupPassword && (
                        <Text {...errorStyling}>
                            Please enter your password
                        </Text>
                    )}
                    {firebaseError &&
                    firebaseError.code === 'auth/weak-password' ? (
                        <Text {...errorStyling}>{firebaseError.message}</Text>
                    ) : null}

                    <Label
                        variant="text.formLabel"
                        htmlFor="signup-confirm-password"
                    >
                        CONFIRM PASSWORD
                    </Label>
                    <Input
                        name="signupConfirmPassword"
                        id="signup-confirm-password"
                        type="password"
                        variant="variants.authInput"
                        ref={register({
                            validate: value =>
                                value === getValues('signupPassword'),
                        })}
                    />
                    {errors.signupConfirmPassword && (
                        <Text {...errorStyling}>
                            Your passwords don&apos;t match
                        </Text>
                    )}

                    <Checkbox
                        name="signupNewsletter"
                        id="signup-newsletter"
                        text="Sign me up for the newsletter"
                    />
                </Box>

                <Button type="submit" mt={[5]} px={[9]} py={[4]}>
                    CREATE ACCOUNT
                </Button>
            </Box>
        </FormProvider>
    );
};

export { SignUp };
