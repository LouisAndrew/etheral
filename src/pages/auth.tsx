import React from 'react';
import { PageProps } from 'gatsby';
import { Layout } from 'components/layout';
import Auth from 'components/auth';

const AuthPage = (props: PageProps) => {
    return (
        <Layout>
            <div className="top">
                <Auth />
            </div>
        </Layout>
    );
};

export default AuthPage;
