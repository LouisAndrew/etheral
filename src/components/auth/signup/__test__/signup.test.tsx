import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { set } from 'lodash';
import userEvent from '@testing-library/user-event';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { SignUp } from '../signup';
import { SignUpProps } from 'components/auth/auth';

describe('Signup component', () => {
    const signupDataContainer = {
        email: '',
        password: '',
        name: '',
    };
    const signup = jest.fn(({ email, password, name }: SignUpProps) => {
        set(signupDataContainer, 'email', email);
        set(signupDataContainer, 'password', password);
        set(signupDataContainer, 'name', name);
    });

    const el: React.ReactElement = <SignUp signup={signup} />;

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('should render and submits correctly', () => {
        const { getByLabelText, getByRole } = render(el);

        const emailInput: HTMLElement = getByLabelText('EMAIL');
        const passwordInput: HTMLElement = getByLabelText('PASSWORD');
        const confirmPassword: HTMLElement = getByLabelText('CONFIRM PASSWORD');
        const nameInput: HTMLElement = getByLabelText('FIRST NAME');

        const email = 'aa@aa';
        const password = '124';
        const name = 'Alice';

        userEvent.type(emailInput, email);
        userEvent.type(passwordInput, password);
        userEvent.type(confirmPassword, password);
        userEvent.type(nameInput, name);

        const submitButton = getByRole('button', { name: 'CREATE ACCOUNT' });
        userEvent.click(submitButton);

        setTimeout(() => {
            expect(signupDataContainer.email).toBe(email);
            expect(signupDataContainer.password).toBe(password);
            expect(signupDataContainer.name).toBe(name);
        }, 800);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
