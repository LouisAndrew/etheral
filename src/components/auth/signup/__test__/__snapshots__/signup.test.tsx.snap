// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component matches snapshot 1`] = `
<form
  className="css-1lx7oj"
  onSubmit={[Function]}
>
  <h3
    className="css-d2e41n"
  >
    NEW ACCOUNT
  </h3>
  <div
    className="css-vurnku"
  >
    <label
      className="css-knfhfe"
      htmlFor="signup-first-name"
    >
      FIRST NAME
    </label>
    <input
      className="css-4y926w"
      id="signup-first-name"
      name="signupFirstName"
      type="text"
    />
    <label
      className="css-knfhfe"
      htmlFor="signup-last-name"
    >
      LAST NAME
    </label>
    <input
      className="css-4y926w"
      id="signup-last-name"
      name="signupLastName"
      type="text"
    />
    <label
      className="css-knfhfe"
      htmlFor="signup-email"
    >
      EMAIL
    </label>
    <input
      className="css-4y926w"
      id="signup-email"
      name="signupEmail"
      type="email"
    />
    <label
      className="css-knfhfe"
      htmlFor="signup-password"
    >
      PASSWORD
    </label>
    <input
      className="css-4y926w"
      id="signup-password"
      name="signupPassword"
      type="password"
    />
    <label
      className="css-knfhfe"
      htmlFor="signup-confirm-password"
    >
      CONFIRM PASSWORD
    </label>
    <input
      className="css-4y926w"
      id="signup-confirm-password"
      name="signupConfirmPassword"
      type="password"
    />
    <input
      className="css-34q1wz"
      id="signup-newsletter"
      name="signupNewsletter"
      type="checkbox"
    />
    <label
      className="css-7tv9me"
      htmlFor="signup-newsletter"
    >
      <div
        className="css-4cffwv"
        id="newsletter-checkbox"
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M10 15.172l9.192-9.193l1.415 1.414L10 18l-6.364-6.364l1.414-1.414z\\" fill=\\"#fff\\"/>",
            }
          }
          focusable={false}
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "transform": "rotate(360deg)",
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
      </div>
      Sign me up for the newsletter
    </label>
  </div>
  <button
    className="css-1kygmdx"
    type="submit"
  >
    CREATE ACCOUNT
  </button>
</form>
`;
